src/v/cloud_storage_clients/configuration.cc

```diff
@@ -190,25 +194,61 @@ model::cloud_storage_backend infer_backend_from_configuration(
         case model::cloud_credentials_source::aws_instance_metadata:
             [[fallthrough]];
         case model::cloud_credentials_source::sts:
+            vlog(
+              client_config_log.info,
+              "cloud_storage_backend derived from cloud_credentials_source {} "
+              "as aws",
+              cloud_storage_credentials_source);
             return model::cloud_storage_backend::aws;
         case model::cloud_credentials_source::gcp_instance_metadata:
+            vlog(
+              client_config_log.info,
+              "cloud_storage_backend derived from cloud_credentials_source {} "
+              "as google_s3_compat",
+              cloud_storage_credentials_source);
             return model::cloud_storage_backend::google_s3_compat;
         case model::cloud_credentials_source::config_file:
-            __builtin_unreachable();
+            vassert(
+              false,
+              "unexpected credential source {}",
+              cloud_storage_credentials_source);
             break;
         }
     }
 
     auto& s3_config = std::get<s3_configuration>(client_config);
     const auto& uri = s3_config.uri;
     if (uri().find("google") != uri().npos) {
+        vlog(
+          client_config_log.info,
+          "cloud_storage_backend set to google_s3_compat because uri {} "
+          "matches pattern 'google'",
+          uri());
         return model::cloud_storage_backend::google_s3_compat;
     }
 
+    // For unit tests, allow simulating AWS like behavior
+    if (uri().find("127.0.0.1") != uri().npos) {
+        vlog(
+          client_config_log.info,
+          "cloud_storage_backend set to aws because uri {} "
+          "matches pattern for localhost",
+          uri());
+        return model::cloud_storage_backend::aws;
+    }
+
     if (uri().find("minio") != uri().npos) {
+        vlog(
+          client_config_log.info,
+          "cloud_storage_backend set to aws because uri {} "
+          "matches pattern for minio",
+          uri());
         return model::cloud_storage_backend::minio;
     }
 
+    vlog(
```

Comment:
Are we missing an AWS case here?  If someone sets a key+secret, but is using `config_file` credentials, and has left uri blank, then `s3_configuration::make_configuration` substitutes the AWS URL (and we shouldn't log "unable to infer")

Reply:
fixed

Reply:
I have added a match expression helper to the string switch to match urls to enum types